lua/entities/gmod_wire_characterlcd/cl_init.lua

```diff
@@ -0,0 +1,614 @@
+  Address = math.floor(Address)
+  if Address < 0 then return nil end
+  if Address >= 1024 then return nil end
+
+  return self.Memory2[Address]
+end
+
+
+
+function ENT:ShiftScreenRight()
+  if self.NewClk then
+    for y=0,self.ScreenHeight-1 do
+      for x=self.ScreenWidth-1,1,-1 do
+        self.Memory1[x+y*self.ScreenWidth] = self.Memory1[x+y*self.ScreenWidth-1]
+      end
+      self.Memory1[y*self.ScreenWidth] = 0
+    end
+  end
+  for y=0,self.ScreenHeight-1 do
+    for x=self.ScreenWidth-1,1,-1 do
+      self.Memory2[x+y*self.ScreenWidth] = self.Memory2[x+y*self.ScreenWidth-1]
+    end
+    self.Memory2[y*self.ScreenWidth] = 0
+  end
+end
+
+function ENT:ShiftScreenLeft()
+  if self.NewClk then
+    for y=0,self.ScreenHeight-1 do
+      for x=0,self.ScreenWidth-2 do
+      self.Memory1[x+y*self.ScreenWidth] = self.Memory1[x+y*self.ScreenWidth+1]
+      end
+    self.Memory1[y*self.ScreenWidth+self.ScreenWidth-1] = 0
+    end
+  end
+  for y=0,self.ScreenHeight-1 do
+    for x=0,self.ScreenWidth-2 do
+    self.Memory2[x+y*self.ScreenWidth] = self.Memory2[x+y*self.ScreenWidth+1]
+    end
+  self.Memory2[y*self.ScreenWidth+self.ScreenWidth-1] = 0
+  end
+end
+
+function ENT:WriteCell(Address,value)
+  Address = math.floor(Address)
+  if Address < 0 then return false end
+  if Address >= 1024 then return false end
+
+  if Address == 1023 then self.NewClk = value ~= 0 end
+  if Address == 1009 and (value*self.Memory2[1010] > 1003 or value*18 > 1024) then return false end
+  if Address == 1010 and (value*self.Memory2[1009] > 1003 or value*24 > 1024) then return false end
+  if self.NewClk then
+    self.Memory1[Address] = value -- Vis mem
+    self.NeedRefresh = true
+  end
+  self.Memory2[Address] = value -- Invis mem
+
+  if Address == 1011 then
+    
+    if self.Memory1[1015] >= 1 then
+      if self.Memory1[1014] >= 1 then
+        self:ShiftScreenRight()
+      else
+        self:ShiftScreenLeft()
+      end
+      self.Memory1[self.Memory1[1021]] = value
+      self.Memory2[self.Memory1[1021]] = value
+    else
+      self.Memory1[self.Memory1[1021]] = value
+      self.Memory2[self.Memory1[1021]] = value
+      if self.Memory1[1014] >= 1 then
+        self.Memory1[1021] = (self.Memory1[1021] - 1)%(self.ScreenHeight*self.ScreenWidth)
+      else
+        self.Memory1[1021] = (self.Memory1[1021] + 1)%(self.ScreenHeight*self.ScreenWidth)
+      end
+      self.Memory2[1021] = self.Memory1[1021]
+    end
+    
+  end
+  if Address == 1017 then
+    for i = 0, self.ScreenWidth-1 do
+      self.Memory1[value*self.ScreenWidth+i] = 0
+      self.Memory2[value*self.ScreenWidth+i] = 0
+    end
+    self.NeedRefresh = true
+  end
+  if Address == 1018 then
+    for i = 0, self.ScreenWidth*self.ScreenHeight-1 do
+      self.Memory1[i] = 0
+      self.Memory2[i] = 0
+    end
+    self.NeedRefresh = true
+  end
+  if Address == 1009 then
+    self.ScreenWidth = value
+  end
+  if Address == 1010 then
+    self.ScreenHeight = value
+  end
```

Comment:
Values like these need to be clamped to prevent huge values causing infinite loops or crashes to clients.

Reply:
Ah I see. Could you move those checks into these blocks? It would also be good if you used elseif's for the address checks.

Reply:
If that check was in the blocks then Memory2 would have it set.

Reply:
It also looks like you can use address 1021 and 1011 to inject whatever value you want anywhere. For address 1017, value also needs to be sanitized.

Reply:
I'm also not sure why Memory2 is needed. Wouldn't something like this work fine?

```lua
function ENT:WriteCell(Address,value)
	Address = math.floor(Address)
	if Address < 0 or Address >= 1024 then return false end

	if Address == 1009 then
		if value*self.Memory[1010] > 1003 or value*18 > 1024 then return false end
		self.ScreenWidth = value
	elseif Address == 1010 then
		if value*self.Memory[1009] > 1003 or value*24 > 1024 then return false end
		self.ScreenHeight = value
	elseif Address == 1011 then
		if self.Memory[1015] >= 1 then
			if self.Memory[1014] >= 1 then
				self:ShiftScreenRight()
			else
				self:ShiftScreenLeft()
			end
			self.Memory[self.Memory[1021]] = value
		else
			self.Memory[self.Memory[1021]] = value
			if self.Memory[1014] >= 1 then
				self.Memory[1021] = (self.Memory[1021] - 1)%(self.ScreenHeight*self.ScreenWidth)
			else
				self.Memory[1021] = (self.Memory[1021] + 1)%(self.ScreenHeight*self.ScreenWidth)
			end
		end
	elseif Address == 1017 then
		for i = 0, self.ScreenWidth-1 do
			self.Memory[value*self.ScreenWidth+i] = 0
		end
		self.NeedRefresh = true
	elseif Address == 1018 then
		for i = 0, self.ScreenWidth*self.ScreenHeight-1 do
			self.Memory[i] = 0
		end
		self.NeedRefresh = true
	elseif Address == 1023 and value~=0 then
		self.NeedRefresh = true
	end

	self.Memory[Address] = value

	return true
end
```

Reply:
Should also probably change the 'self.ScreenWidth' and 'self.ScreenHeight' vars into 'self.Memory[1009]' and 'self.Memory[1010]'

Reply:
Also maybe change self.NeedRefresh to self.Memory[1023] and reset that to 0 after refreshing.